### PR TITLE
Add NetEase provider

### PR DESCRIPTION
**Pull request by HahaMango**
Hi, I implemented a NetEase provider, also add the test code.

Hot to use:

```csharp
builder.AddNetEase(options => 
{
    options.ClientId = "your client Id ";
    options.ClientSecret = "your client secret ";
});
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,6 +137,7 @@ If a provider you're looking for does not exist, consider making a PR to add one
...
 | Myob | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.Myob?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.Myob/ "Download AspNet.Security.OAuth.Myob from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.Myob?includePreReleases=false)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Myob "Download AspNet.Security.OAuth.Myob from MyGet.org") | [Documentation](https://developer.myob.com/api/accountright/api-overview/authentication/ "Myob developer documentation") |
+|  NetEase  | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.NetEase?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.NetEase/ "Download AspNet.Security.OAuth.NetEase from NuGet.org") |  [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.NetEase?includePreReleases=false)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.NetEase "Download AspNet.Security.OAuth.NetEase from MyGet.org") |  [Documentation](http://reg.163.com/help/help_oauth2.html "NetEase developer documentation") |
```
Comment (martincostello) [inline README.md]:
```suggestion
| NetEase | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.NetEase?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.NetEase/ "Download AspNet.Security.OAuth.NetEase from NuGet.org") |  [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.NetEase?includePreReleases=false)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.NetEase "Download AspNet.Security.OAuth.NetEase from MyGet.org") |  [Documentation](https://reg.163.com/help/help_oauth2.html "NetEase developer documentation") |
```

File: src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationDefaults.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="RemoteAuthenticationOptions.CallbackPath"/>.
+        /// </summary>
+        public const string CallbackPath = "/signin-netease";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "http://reg.163.com/open/oauth2/authorize.do";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationDefaults.cs]:
```suggestion
        public const string AuthorizationEndpoint = "https://reg.163.com/open/oauth2/authorize.do";
```

File: src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "http://reg.163.com/open/oauth2/token.do";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationDefaults.cs]:
```suggestion
        public const string TokenEndpoint = "https://reg.163.com/open/oauth2/token.do";
```

File: src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationOptions.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            ClaimsIssuer = NetEaseAuthenticationDefaults.Issuer;
+
+            CallbackPath = NetEaseAuthenticationDefaults.CallbackPath;
+
+            AuthorizationEndpoint = NetEaseAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = NetEaseAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = NetEaseAuthenticationDefaults.UserInformationEndpoint;
+
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "userId");
+            ClaimActions.MapJsonKey(ClaimTypes.Name, "username");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationOptions.cs:29]: User name isn't really a name if it's also an email address.

File: test/AspNet.Security.OAuth.Providers.Tests/NetEase/NetEaseTests.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+        protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
+        {
+            builder.AddNetEase(options => ConfigureDefaults(builder, options));
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.NameIdentifier, "820014421")]
+        [InlineData(ClaimTypes.Name, "urstest_mreg")]
+        [InlineData(ClaimTypes.Email, "urstest_mreg")]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/NetEase/NetEaseTests.cs]: This doesn't look like an email address?

File: test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json [hunk 6]
```diff
@@ -0,0 +1,24 @@
+{
+  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/master/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
+  "items": [
+    {
+      "comment": "http://reg.163.com/help/help_oauth2.html",
+      "uri": "http://reg.163.com/open/oauth2/token.do",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json]:
```suggestion
      "uri": "https://reg.163.com/open/oauth2/token.do",
```

File: test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{
+  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/master/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
+  "items": [
+    {
+      "comment": "http://reg.163.com/help/help_oauth2.html",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json]:
```suggestion
      "comment": "https://reg.163.com/help/help_oauth2.html",
```

File: test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      "uri": "http://reg.163.com/open/oauth2/token.do",
+      "method": "POST",
+      "contentFormat": "json",
+      "contentJson": {
+        "access_token": "secret-access-token",
+        "expires_in": "5184000"
+      }
+    },
+    {
+      "comment": "See http://reg.163.com/help/help_oauth2.html",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/NetEase/bundle.json]:
```suggestion
      "comment": "See https://reg.163.com/help/help_oauth2.html",
```

File: src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationOptions.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+            ClaimActions.MapJsonKey(ClaimTypes.Email, "username");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.NetEase/NetEaseAuthenticationOptions.cs]: The user name, at least in the sample in the docs, isn't an email address.

**Review by martincostello — CHANGES_REQUESTED** [review]
As an OAuth provider, HTTPS needs to be used everywhere.

**Comment by HahaMango** [comment]
@martincostello Yes, according to the document it's just a user name, not a mailbox, my fault. I've changed it.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks again for your contribution @HahaMango 🎉

**Comment by martincostello** [comment]
This provider is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.NetEase/3.1.0

**Pull request merged**